pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,508 @@
+				}
+
+				k.waitNextMu.Lock()
+				if next, ok := k.waitNext[peer.String()]; ok && time.Now().Before(next) {
+					k.waitNextMu.Unlock()
+					return false, false, nil
+				}
+				k.waitNextMu.Unlock()
+
+				currentDepth := k.NeighborhoodDepth()
+				if k.saturationFunc(po, currentDepth, k.connectedPeers) {
+					return false, true, nil // bin is saturated, skip to next bin
+				}
+
+				ma, err := k.addressBook.Get(peer)
+				if err != nil {
+					// either a peer is not known in the address book, in which case it
+					// should be removed, or that some severe I/O problem is at hand
+
+					if errors.Is(err, storage.ErrNotFound) {
+						k.logger.Errorf("failed to get address book entry for peer: %s", peer.String())
+						peerToRemove = peer
+						return false, false, errMissingAddressBookEntry
+					}
+
+					return false, false, err
+				}
+				k.logger.Debugf("kademlia dialing to peer %s", peer.String())
+
+				err = k.connect(ctx, peer, ma, po)
+				if err != nil {
+					k.logger.Errorf("error connecting to peer %s: %v", peer, err)
+					// continue to next
+					return false, false, nil
+				}
+
+				k.waitNextMu.Lock()
+				delete(k.waitNext, peer.String())
+				k.waitNextMu.Unlock()
+
+				k.depthMu.Lock()
+				k.depth = k.recalcDepth()
+				k.depthMu.Unlock()
+
+				k.logger.Debugf("connected to peer: %s old depth: %d new depth: %d", peer, currentDepth, k.NeighborhoodDepth())
+
+				select {
+				case <-k.quit:
+					// shutting down
+					return true, false, nil
+				default:
+				}
+
+				// the bin could be saturated or not, so a decision cannot
+				// be made before checking the next peer, so we iterate to next
+				return false, false, nil
+			})
+
+			if err != nil {
+				if errors.Is(err, errMissingAddressBookEntry) {
+					po := uint8(swarm.Proximity(k.base.Bytes(), peerToRemove.Bytes()))
+					k.knownPeers.Remove(peerToRemove, po)
+				} else {
+					k.logger.Errorf("kademlia manage loop iterator: %v", err)
+				}
+			}
+		}
+	}
+}
+
+// binSaturated indicates whether a certain bin is saturated or not.
+// when a bin is not saturated it means we would like to proactively
+// initiate connections to other peers in the bin.
+func binSaturated(bin, depth uint8, peers *pslice.PSlice) bool {
+	// short circuit for bins which are >= depth
+	if bin >= depth {
+		return false
+	}
+
+	// lets assume for now that the minimum number of peers in a bin
+	// would be 2, under which we would always want to connect to new peers
+	// obviously this should be replaced with a better optimization
+	// the iterator is used here since when we check if a bin is saturated,
+	// the plain number of size of bin might not suffice (for example for squared
+	// gaps measurement)
+
+	size := 0
+	_ = peers.EachBin(func(_ swarm.Address, po uint8) (bool, bool, error) {
+		if po == bin {
+			size++
+		}
+		return false, false, nil
+	})
+
+	return size >= 2
+}
+
+// recalcDepth returns the kademlia depth. Must be called under lock.
```

Comment:
Does this mean that it is not thread-safe and should be called under outside lock? If so, I would just  say something like "It is not safe to use concurentlly"  instead "must be called under lock". Also, if it is needed to be called concurentlly, I would export a thread safe version as well as it should be easy to do.

Reply:
:+1: I changed the comment. it actually is not crucial for the function itself to be under lock, since the pslice structs handle locking themselves, so it is not critical.